File: projects/composablekernel/python/ck_dsl_c/tests/parity/fused_moe_e2e_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/fused_moe_e2e_emit.py -- Python reference emitter for the
# fused_moe_e2e (end-to-end fused-MoE forward orchestrator) parity harness.
#
# Selects one of the sampled FusedMoeForwardSpec configs by argv[1] (the config
# index), constructs FusedMoeForward(spec) so the __init__ shape-aware tile-swap
# policy + arch resolution runs exactly as the live orchestrator would, then
# lowers each lowerable pipeline stage to AMDGPU LLVM IR via lower_kernel_to_llvm
# and prints them concatenated to stdout, each prefixed with a stage banner, so
# the output can be byte-compared with the C emitter fused_moe_e2e_emit.c.
#
# The orchestrator emits NO single monolithic kernel; it composes pre-existing
# sub-kernel builders selected by the (tile-policy-adjusted) spec. The C port's
# ckc_fused_moe_forward_lower_to_llvm currently wires three lowerable stages --
# ROUTER (topk_softmax), GATE_UP_GEMM and DOWN_GEMM (both the batched-GEMM
# builder shape) -- so this emitter lowers exactly those three, in the same
# order, to keep the byte comparison apples-to-apples. (The sort / gather /
# silu_mul / topk_reduce stages return NOTIMPL on the C side and are therefore
# excluded from the diff here.)
import sys

from ck_dsl import lower_kernel_to_llvm
from ck_dsl.instances.common.fused_moe_e2e import (
    FusedMoeForwardSpec,
    FusedMoeForward,
)
from ck_dsl.instances.common.topk_softmax import build_topk_softmax
from ck_dsl.instances.common.batched_gemm import build_batched_gemm


# The six sampled configs (must match the C emitter's make_spec()).
_CONFIGS = (
    dict(tokens=1, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=8, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=32, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=128, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=1, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="bf16"),
    dict(tokens=128, experts=32, topk=5, hidden=8192, intermediate=8192, dtype="f16"),
)


def _spec(idx: int) -> FusedMoeForwardSpec:
    if idx < 0 or idx >= len(_CONFIGS):
        raise SystemExit(f"unknown config index {idx}")
    cfg = _CONFIGS[idx]
    return FusedMoeForwardSpec(arch="gfx950", **cfg)


# Lowerable stages, in the C emitter's order. Each entry is (banner, builder)
# where the builder takes the (tile-policy-adjusted) FusedMoeForward instance.
def _emit_router(fwd: FusedMoeForward) -> str:
    spec = fwd.spec.to_topk_softmax_spec()
    kernel = build_topk_softmax(spec)
    return lower_kernel_to_llvm(kernel, arch="gfx950")


def _emit_batched_gemm(fwd: FusedMoeForward) -> str:
    # GATE_UP_GEMM and DOWN_GEMM both lower via the batched-GEMM builder shape
    # (the orchestrator parameterises it per-call; the representative .ll for
    # golden diffing is the batched_gemm builder), matching the C port.
    spec = fwd.spec.to_batched_gemm_spec()
    kernel = build_batched_gemm(spec)
    return lower_kernel_to_llvm(kernel, arch="gfx950")


_STAGES = (
    ("ROUTER", _emit_router),
    ("GATE_UP_GEMM", _emit_batched_gemm),
    ("DOWN_GEMM", _emit_batched_gemm),
)


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write("usage: fused_moe_e2e_emit.py <config_index 0..5>\n")
        return 2
    idx = int(sys.argv[1])
    spec = _spec(idx)
    # Construct the orchestrator so __init__ runs the arch resolve + shape-aware
    # tile-swap policy; the stages below read the post-swap fwd.spec.
    fwd = FusedMoeForward(spec)
    out = []
    for banner, emit in _STAGES:
        out.append(f"; === fused_moe_e2e stage: {banner} ===\n")
        out.append(emit(fwd))
        if not out[-1].endswith("\n"):
            out.append("\n")
    sys.stdout.write("".join(out))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: projects/composablekernel/python/ck_dsl_c/tests/parity/fused_moe_e2e_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/fused_moe_e2e_emit.c -- C-side emitter for the fused_moe_e2e
 * (end-to-end fused-MoE forward orchestrator) parity harness.
 *
 * Selects one of the sampled FusedMoeForwardSpec configs by argv[1] (the config
 * index), materialises the spec via ckc_fmoe_forward_spec_default() + the per-
 * config shape fields, then lowers each lowerable pipeline stage to AMDGPU LLVM
 * IR text via ckc_fused_moe_forward_lower_to_llvm (arch gfx950, flavor AUTO) and
 * prints them concatenated to stdout, each prefixed with a stage banner, so the
 * output can be byte-compared with the Python emitter fused_moe_e2e_emit.py.
 *
 * The orchestrator emits NO single monolithic kernel; ckc_fused_moe_forward_
 * lower_to_llvm runs __init__'s arch resolve + tile-swap policy internally (via
 * the build ctx) and delegates to the spec-selected sub-kernel builder. The
 * three lowerable stages are ROUTER (topk_softmax) and GATE_UP_GEMM / DOWN_GEMM
 * (both the batched-GEMM builder shape). The remaining stages (sort hist/scan/
 * scatter, gather, silu_mul, topk_reduce) are TODO(port) NOTIMPL and are
 * therefore excluded here, matching the Python emitter.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ckc/ir.h"
#include "ckc/lower_llvm.h"
#include "ckc/instance_fused_moe_e2e.h"

/* Populate the FusedMoeForwardSpec for config `idx`. Returns 0, or -1 on an
 * unknown index. Every non-shape field stays at the dataclass default
 * (ckc_fmoe_forward_spec_default), so the tile-swap policy and the static gate
 * see exactly the Python defaults; only the enumerated shape + dtype differ. */
static int make_spec(int idx, ckc_fmoe_forward_spec_t *spec) {
    *spec = ckc_fmoe_forward_spec_default();
    spec->arch = "gfx950";
    switch (idx) {
    case 0: /* tokens=1   E8  K2 H4096 I7168 f16  */
        spec->tokens = 1;   spec->experts = 8;  spec->topk = 2;
        spec->hidden = 4096; spec->intermediate = 7168; spec->dtype = "f16";
        break;
    case 1: /* tokens=8   E8  K2 H4096 I7168 f16  */
        spec->tokens = 8;   spec->experts = 8;  spec->topk = 2;
        spec->hidden = 4096; spec->intermediate = 7168; spec->dtype = "f16";
        break;
    case 2: /* tokens=32  E8  K2 H4096 I7168 f16  */
        spec->tokens = 32;  spec->experts = 8;  spec->topk = 2;
        spec->hidden = 4096; spec->intermediate = 7168; spec->dtype = "f16";
        break;
    case 3: /* tokens=128 E8  K2 H4096 I7168 f16  */
        spec->tokens = 128; spec->experts = 8;  spec->topk = 2;
        spec->hidden = 4096; spec->intermediate = 7168; spec->dtype = "f16";
        break;
    case 4: /* tokens=1   E8  K2 H4096 I7168 bf16 */
        spec->tokens = 1;   spec->experts = 8;  spec->topk = 2;
        spec->hidden = 4096; spec->intermediate = 7168; spec->dtype = "bf16";
        break;
    case 5: /* tokens=128 E32 K5 H8192 I8192 f16  */
        spec->tokens = 128; spec->experts = 32; spec->topk = 5;
        spec->hidden = 8192; spec->intermediate = 8192; spec->dtype = "f16";
        break;
    default:
        return -1;
    }
    return 0;
}

/* Lowerable stages, in the Python emitter's order. GATE_UP_GEMM and DOWN_GEMM
 * both resolve to the batched-GEMM builder shape inside
 * ckc_fused_moe_forward_lower_to_llvm. */
static const struct { const char *banner; ckc_fmoe_stage_t stage; } STAGES[] = {
    {"ROUTER",       CKC_FMOE_STAGE_ROUTER},
    {"GATE_UP_GEMM", CKC_FMOE_STAGE_GATE_UP_GEMM},
    {"DOWN_GEMM",    CKC_FMOE_STAGE_DOWN_GEMM},
};

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <config_index 0..5>\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);

    ckc_fmoe_forward_spec_t spec;
    if (make_spec(idx, &spec) != 0) {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    const size_t nstages = sizeof(STAGES) / sizeof(STAGES[0]);
    for (size_t i = 0; i < nstages; i++) {
        char *llvm_text = NULL;
        char err[CKC_ERR_MSG_CAP];
        err[0] = 0;
        ckc_status_t st = ckc_fused_moe_forward_lower_to_llvm(
            &spec, "gfx950", STAGES[i].stage, CKC_LLVM_FLAVOR_AUTO,
            &llvm_text, err, sizeof err);
        if (st != CKC_OK || !llvm_text) {
            fprintf(stderr, "lower failed (config %d stage %s): status=%d err=%s\n",
                    idx, STAGES[i].banner, (int)st, err);
            free(llvm_text);
            return 1;
        }
        printf("; === fused_moe_e2e stage: %s ===\n", STAGES[i].banner);
        fputs(llvm_text, stdout);
        size_t n = strlen(llvm_text);
        if (n == 0 || llvm_text[n - 1] != '\n') {
            fputc('\n', stdout);
        }
        free(llvm_text);
    }
    return 0;
}
